### PR TITLE
Automatically determine oPower subdomain for PEPCO and Delmarva

### LINKED ISSUE #14: Subdomains for PEPCO and Delmarva Power are incorrect for some states
Hi,

I was trying to setup the Home Assistant plugin and kept getting this error (not sure if related or separate issue):

```yml
2023-08-02 22:02:58.615 ERROR (MainThread) [homeassistant.components.opower.coordinator] Unexpected error fetching Opower data: 'NoneType' object has no attribute 'group'
Traceback (most recent call last):
File "/lsiopy/lib/python3.11/site-packages/homeassistant/helpers/update_coordinator.py", line 283, in _async_refresh
self.data = await self._async_update_data()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/lsiopy/lib/python3.11/site-packages/homeassistant/components/opower/coordinator.py", line 67, in _async_update_data
await self.api.async_login()
File "/lsiopy/lib/python3.11/site-packages/opower/opower.py", line 151, in async_login
self.access_token = await self.utility.async_login(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/lsiopy/lib/python3.11/site-packages/opower/utilities/exelon.py", line 44, in async_login
settings = json.loads(re.search(r"var SETTINGS = ({.*});", result).group(1))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'group'
```

So instead I tried to use this python package directly outside of HA, which gave me the error:

```
raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 403, message='Forbidden', url=URL('https://pep.opower.com/ei/edge/apis/multi-account-v1/cws/pep/customers?offset=0&batchSize=100&addressFilter=')
```

When loading https://secure.pepco.com/api/Services/MyAccountService.svc/GetConfiguration I can see the following values

image

When logging into my account, all requests seem to prefer `oPowerURLBaseJurisdiction`:

image

I was able to get the python script working by [changing this](https://github.com/tronikos/opower/blob/8d7f2a99e313c07a1975b3a50f40b2c950ac7d84/src/opower/utilities/pepco.py#L18C16-L18C21) to `pepd`.

I am unsure if this is dynamic, as it seemed to work when it was initially added in #3

### PR DESCRIPTION
**Pull request by trancefam**
Fixes #14 

Add functionality to automatically determine subdomain in `exelon.py` base class. Exelon utilities that do not need this override it in their derived class. Those that do need it (PEPCO and Delmarva), use the default base class functionality.

I have tested & confirmed this works with PEPCO.

### CODE REVIEW

**Comment by ronaldheft** [comment]
I can give this a review tomorrow and ensure it works for Pepco MD.

File: src/opower/utilities/exelon.py [hunk 1]
```diff
@@ -25,11 +32,63 @@ def timezone() -> str:
...
+                #      we default to taking the first in the list. Future enhancement is to support
+                #      multiple accounts (which could result in different subdomain for each)
+                active_accounts = [account for account in response['data'] if account['status'] == 'Active']
+                isResidential = active_accounts[0]["isResidential"]
+                state = active_accounts[0]['PremiseInfo'][0]['mainAddress']['townDetail']['stateOrProvince']
+                _LOGGER.debug("found exelon account isResidential: %s", isResidential)
+                _LOGGER.debug("found exelon account state: %s", state)
+
+            # Determine subdomain to use by matching logic found in https://cls.login_domain()/dist/app.js
+            Exelon._subdomain = oPowerURLBase
```
Comment (tronikos) [inline src/opower/utilities/exelon.py:77]: Can you initialize this in async_login instead and keep the subdomain method synchronous?

File: src/opower/utilities/exelon.py [hunk 2]
```diff
@@ -25,11 +32,63 @@ def timezone() -> str:
...
+    
+    @classmethod
+    async def subdomain(cls) -> str:
+        """Return the opower.com subdomain for this utility."""
+        if Exelon._subdomain is not None:
+            return Exelon._subdomain
+        else:
+            # Get the potential url subdomains, they can vary based on account type/location
+            async with cls._session.get(
+                "https://" + cls.login_domain() + "/api/Services/MyAccountService.svc/GetConfiguration",
```
Comment (tronikos) [inline src/opower/utilities/exelon.py:44]: I'd avoid this call and instead have a method that returns the secondary subdomain that defaults to None except for PEPCO and Delmarva.

File: src/opower/utilities/exelon.py [hunk 3]
```diff
@@ -2,16 +2,23 @@
...
 from ..exceptions import InvalidAuth
 
+_LOGGER = logging.getLogger(__file__)
+
 
 class Exelon:
     """Base class for Exelon subsidiaries."""
 
+    _access_token = None
+    _subdomain = None
```
Comment (tronikos) [inline src/opower/utilities/exelon.py:19]: Having the domain here is fine for now, Evergy is doing the same, but let's not add any more state in the class like access token and session. At some I want to refactor all these to be normal, non-static classes to allow adding multiple config entries in Home Assistant for the same utility but potentially different subdomain. Hopefully there shouldn't be anyone affected by this limitation, but if you add access token and session, that means people cannot have multiple config entries for the same utility and same domain which I imagine some people might be affected, e.g. with rentals in the same location.

**Review by tronikos — CHANGES_REQUESTED** [review]

**Comment by ronaldheft** [comment]
Functionality-wise, this worked great with my Pepco MD account. My old DC address is not the active account and is 4+ years old, so I was not able to test with a DC address.

File: src/opower/utilities/exelon.py (hunk 2 above)
Comment (trancefam) [reply src/opower/utilities/exelon.py]: removed call to `GetConfiguration`

**Review by trancefam — COMMENTED** [review]
Updated with suggestions

**Review by tronikos — APPROVED** [review]

**Comment by tronikos** [comment]
Can you please confirm this works correctly after https://github.com/tronikos/opower/pull/25?

**Comment by trancefam** [comment]
Yes it does.

**Pull request merged**